core/lib/zksync_core/src/sync_layer/batch_status_updater/mod.rs

```diff
@@ -0,0 +1,451 @@
+/// node on these batches in order to see whether the status has changed. If some changes were picked up,
+/// the module updates the database to mirror the state observable from the main node.
+#[derive(Debug)]
+pub struct BatchStatusUpdater {
+    client: Box<dyn MainNodeClient>,
+    pool: ConnectionPool,
+    sleep_interval: Duration,
+    changes_sender: mpsc::UnboundedSender<StatusChanges>,
+}
+
+impl BatchStatusUpdater {
+    const DEFAULT_SLEEP_INTERVAL: Duration = Duration::from_secs(5);
+
+    pub fn new(main_node_url: &str, pool: ConnectionPool) -> anyhow::Result<Self> {
+        let client = HttpClientBuilder::default()
+            .build(main_node_url)
+            .context("Unable to create a main node client")?;
+        Ok(Self::from_parts(
+            Box::new(client),
+            pool,
+            Self::DEFAULT_SLEEP_INTERVAL,
+        ))
+    }
+
+    fn from_parts(
+        client: Box<dyn MainNodeClient>,
+        pool: ConnectionPool,
+        sleep_interval: Duration,
+    ) -> Self {
+        Self {
+            client,
+            pool,
+            sleep_interval,
+            changes_sender: mpsc::unbounded_channel().0,
+        }
+    }
+
+    pub async fn run(self, stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        let mut storage = self.pool.access_storage_tagged("sync_layer").await?;
+        let mut cursor = UpdaterCursor::new(&mut storage).await?;
+        drop(storage);
+        tracing::info!("Initialized batch status updater cursor: {cursor:?}");
+
+        loop {
+            if *stop_receiver.borrow() {
+                tracing::info!("Stop signal received, exiting the batch status updater routine");
+                return Ok(());
+            }
+
+            // Status changes are created externally, so that even if we will receive a network error
+            // while requesting the changes, we will be able to process what we already fetched.
+            let mut status_changes = StatusChanges::default();
+            // Note that we don't update `cursor` here (it is copied), but rather only in `apply_status_changes`.
+            match self.get_status_changes(&mut status_changes, cursor).await {
+                Ok(()) => { /* everything went smoothly */ }
+                Err(UpdaterError::Web3(err)) => {
+                    tracing::warn!("Failed to get status changes from the main node: {err}");
+                }
+                Err(UpdaterError::Internal(err)) => return Err(err),
+            }
+
+            if status_changes.is_empty() {
+                tokio::time::sleep(self.sleep_interval).await;
+            } else {
+                self.apply_status_changes(&mut cursor, status_changes)
+                    .await?;
+            }
+        }
+    }
+
+    /// Goes through the already fetched batches trying to update their statuses.
+    /// Returns a collection of the status updates grouped by the operation type.
```

Comment:
nit: part of the comment starting with "Returns" is a bit misleading, it does not return anything

Reply:
nit: and btw, do you see any specific reason the status_changes are not created inside get_status_changes and returned?

Reply:
`changes` are provided as a mutable reference so that they can be recovered and applied even if a transient error occurs; see the comments in `run()` method.